### DISCUSSION TITLE
Cannot download file in production.

### DISCUSSION THREAD

**Discussion by shadcngeek** [vercel/next.js#61919] [App Router]
I am currently working on a project that involves downloading a folder as zip.  The download woks perfectly in development, but does does not work in production. If I click the download button in production, I get server error with status 500.

It seems like the server is not able to locate folder.

Please help.

```
// app/api/download/route.ts
import { NextRequest, NextResponse } from "next/server";

import fs from "fs";
import path from "path";
import JSZip from "jszip";

export async function POST(request: NextRequest) {
  const { folderToZip } = await request.json();
  const zip = new JSZip();

  await zipFolder(zip, folderToZip, "");
  const zipContent = await zip.generateAsync({ type: "nodebuffer" });

  return new NextResponse(zipContent, {
    status: 200,
    headers: new Headers({
      "content-type": "application/zip",
      "content-disposition": "attachment filename=sample-code.zip",
    }),
  });
}

async function zipFolder(zip: JSZip, folderPath: string, relativePath: string) {
  const files = fs.readdirSync(folderPath);

  for (const file of files) {
    const filePath = path.join(folderPath, file);
    let relativeFilePath: any;
    if (!fs.statSync(filePath).isDirectory()) {
      relativeFilePath = file;
    }

    if (fs.statSync(filePath).isDirectory()) {
      await zipFolder((zip as any).folder(file), filePath, relativeFilePath);
    } else {
      const fileContent = fs.readFileSync(filePath);
      zip.file(relativeFilePath, fileContent);
    }
  }
}

```

```
//actions/handleDownloadZip.ts
const siteUrl =
  process.env.NODE_ENV === "production"
    ? "https://mysite.vercel.app"
    : "http://localhost:3000";

const handleGetFileName = (pathName: string) => {
  const fileName = pathName.split("/").pop();
  return fileName;
};

const handleDownloadZip = async () => {
  try {
    const response = await fetch(`${siteUrl}/api/download`, {
      headers: { "content-type": "application/zip" },
      method: "POST",
      body: JSON.stringify({ folderToZip: "components/cards/activities" }),
    });
    const blob = await response.blob();
    // Create a temporary URL and initiate the download
    const url = window.URL.createObjectURL(new Blob([blob]));
    const link = document.createElement("a");
    link.href = url;
    const folderName = handleGetFileName("components/cards/activities");
    link.setAttribute("download", `${folderName}.zip`);
    document.body.appendChild(link);
    link.click();

    // Clean up
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  } catch (error) {
    console.log(error);
  }
};

export default handleDownloadZip;
```

```
"use client";
import React from "react";

import handleDownloadZip from "@/actions/handleDownloadZip";

function Download() {
  return (
      <div >
        <button onClick={() => handleDownloadZip()}>Download</button>
      </div>
  );
}

export default Download;
```

**Comment by icyJoseph** [comment]
Hi,

A bit hard to follow from your code example, but I think you are missing `process.cwd()`.

In Vercel, an other providers for that matter, server side code, may sometimes run in a task, which runs on a different path, not where your project is. In your code it looks like, the folder to zip path is assuming, that it will be run in the project path, which is why you have it as relative e.g., `components/cards/activities`

Try this instead:

```ts
  const { folderToZip } = await request.json();
  const pathToFolderToZip = `${process.cwd()}/${folderToZip}`;
``` 

And use that variable instead.

**Reply by shadcngeek** [reply]
Hi, I appreciate the time you took to answer my question. Thank you! I have tried this method, but I get the following error message. 
```
message: '{\n  "errno": -2,\n  "syscall": "scandir",\n  "code": "ENOENT",\n  "path": "/var/task/components/landing-page/e-commerce"\n}'
```
I suspect that my component folder is not included in the `build`. I have built the project locally to analyze the folder structure, but the components folder is no where to be found in the build folder.

**Reply by icyJoseph** [reply]
Aha, that's weird... it looks like the `process.cwd` is resolving to where the task is running... 

I think at this point, just put a few logs and try to figure out the right path formation, and that the folder you care about is present at all.

**Reply by shadcngeek** [reply]
Great, I will do that. Thanks.
